File: 头七门/中文版/tools/e2e.mjs
import puppeteer from "puppeteer-core";

const BASE = process.env.BASE || "http://127.0.0.1:8793";
const CHROME =
  process.env.CHROME ||
  "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome";

const fails = [];
function ok(name, cond, extra) {
  if (cond) console.log("PASS", name);
  else {
    console.log("FAIL", name, extra || "");
    fails.push(name + (extra ? " " + extra : ""));
  }
}

async function text(page) {
  return page.evaluate(() => document.body.innerText);
}

const errors = [];
const browser = await puppeteer.launch({
  executablePath: CHROME,
  headless: "new",
  args: ["--no-sandbox", "--disable-gpu", "--no-proxy-server", "--proxy-bypass-list=*"]
});
const page = await browser.newPage();
page.on("pageerror", (e) => errors.push(String(e)));
await page.setViewport({ width: 1280, height: 800 });

async function click(sel) {
  await page.waitForSelector(sel, { timeout: 8000 });
  await page.click(sel);
}

async function disabled(sel) {
  return page.$eval(sel, (el) => el.disabled);
}

try {
  await page.goto(BASE + "/index.html", { waitUntil: "domcontentloaded", timeout: 60000 });
  let t = await text(page);
  ok("title-who", t.indexOf("沈晚闸") !== -1);
  ok("title-task", t.indexOf("钥匙") !== -1 && t.indexOf("有效班") !== -1);
  ok("title-lock", t.indexOf("今晚征兆只能管今晚") !== -1);
  ok("title-no-banned",
    t.indexOf("主线") === -1 && t.indexOf("关卡") === -1 && t.indexOf("通关") === -1 &&
    t.indexOf("解锁") === -1 && t.indexOf("线索") === -1 && t.indexOf("源码") === -1 &&
    t.indexOf("本游戏") === -1 && t.indexOf("唯一解") === -1 && t.indexOf("进度条") === -1);
  ok("title-doc-img", t.indexOf("进站") !== -1);

  await click("#btn-enter");
  t = await text(page);
  ok("boot-who", t.indexOf("沈晚闸") !== -1 && t.indexOf("门岗") !== -1);
  ok("boot-person", t.indexOf("吴成山") !== -1);
  ok("boot-job", t.indexOf("钥匙") !== -1 || t.indexOf("有效班") !== -1);
  ok("boot-doc", t.indexOf("对照流程页") !== -1 && t.indexOf("对照礼簿") !== -1);
  ok("boot-dims", t.indexOf("工钱格") !== -1 && t.indexOf("神智") !== -1 && t.indexOf("馆誉") !== -1 && t.indexOf("门缝") !== -1);
  ok("admit-disabled", await disabled("#btn-admit") === true);
  ok("deny-disabled", await disabled("#btn-deny") === true);
  ok("pencil-hidden", await page.$eval("#paper-pencil", (el) => el.style.display === "none"));
  ok("fang-hidden", await page.$eval("#paper-fang", (el) => el.style.display === "none"));

  const flow = await page.$eval("#btn-benqi", (el) => el.innerText);
  const pencilLater = await page.$eval("#paper-pencil", (el) => el.innerText);
  ok("mutex-not-same", flow.indexOf("只接本期") !== -1);

  await click("#btn-doc-flow");
  ok("grouped", await disabled("#btn-admit") === false);
  t = await text(page);
  ok("grouped-hint", t.indexOf("成组") !== -1);

  await click("#btn-admit");
  t = await text(page);
  ok("n1-admit", t.indexOf("成山") !== -1);
  ok("n1-empty-door", t.indexOf("门外无人") !== -1);

  await click("#btn-next");
  t = await text(page);
  ok("n2", t.indexOf("第2晚") !== -1 && t.indexOf("黄叔") !== -1);
  ok("pencil-shown", await page.$eval("#paper-pencil", (el) => el.style.display !== "none"));

  await click("#btn-doc-pencil");
  await click("#btn-doc-flow");
  t = await text(page);
  ok("true-mutex", t.indexOf("对着") !== -1 || t.indexOf("全放") !== -1 || t.indexOf("外人") !== -1, t.slice(0, 400));
  await click("#btn-deny");
  t = await text(page);
  ok("n2-deny-hs", t.indexOf("黄叔") !== -1);

  await click("#btn-next");
  t = await text(page);
  ok("n3", t.indexOf("第3晚") !== -1);
  ok("omen-need", t.indexOf("广播") !== -1);
  await click("#btn-omen");
  t = await text(page);
  ok("omen-n3", t.indexOf("礼簿") !== -1);
  ok("n3-empty-or-person", t.indexOf("门外无人") !== -1 || t.indexOf("门外") !== -1);
  if ((await page.$eval("#silhouette", (el) => el.textContent)).indexOf("无人") !== -1) {
    await click("#btn-next");
  } else {
    await click("#btn-doc-ledger");
    await click("#btn-deny");
    await click("#btn-next");
  }

  t = await text(page);
  ok("n4", t.indexOf("第4晚") !== -1 && t.indexOf("吴桂香") !== -1);
  ok("fang-shown", await page.$eval("#paper-fang", (el) => el.style.display !== "none"));
  const fangText = await page.$eval("#paper-fang", (el) => el.innerText);
  const flowText = await page.$eval(".paper", (el) => el.innerText);
  ok("fang-not-flow-copy", fangText.indexOf("出嫁女") !== -1 && flowText.indexOf("不接外人") !== -1 && fangText.indexOf("不接外人") === -1);

  await click("#btn-omen");
  await click("#btn-doc-fang");
  await click("#btn-admit");
  t = await text(page);
  ok("n4-admit-wgx", t.indexOf("桂香") !== -1);

  await click("#btn-next");
  t = await text(page);
  ok("n5", t.indexOf("第5晚") !== -1 && t.indexOf("外人") !== -1);
  await click("#btn-omen");
  await click("#btn-doc-flow");
  await click("#btn-deny");

  await click("#btn-next");
  t = await text(page);
  ok("n6", t.indexOf("第6晚") !== -1);
  ok("n6-no-wrong-revisit", t.indexOf("门外无人") !== -1, t.slice(0, 300));
  await click("#btn-next");

  t = await text(page);
  ok("n7", t.indexOf("第7晚") !== -1);
  const sil = await page.$eval("#silhouette", (el) => el.textContent);
  if (sil.indexOf("外人") !== -1) {
    await click("#btn-omen");
    await click("#btn-doc-flow");
    await click("#btn-deny");
  }
  await click("#btn-key");
  t = await text(page);
  ok("benqi", t.indexOf("本期") !== -1 && t.indexOf("¥36") !== -1, t.slice(-400));
  ok("benqi-reason", t.indexOf("放进了") !== -1 || t.indexOf("成山") !== -1, t.slice(-400));

  // 未齐交钥匙
  await page.evaluate(() => { try { localStorage.removeItem("touqi-state"); } catch (e) {} });
  await page.goto(BASE + "/index.html", { waitUntil: "domcontentloaded" });
  await click("#btn-enter");
  await click("#btn-doc-flow");
  await click("#btn-admit");
  await click("#btn-key");
  t = await text(page);
  ok("partial-key", t.indexOf("未齐") !== -1 && t.indexOf("有效班") !== -1, t.slice(-300));
  ok("partial-not-benqi", t.indexOf("¥36") === -1);

  // ghost 打回
  await page.evaluate(() => { try { localStorage.removeItem("touqi-state"); } catch (e) {} });
  await page.goto(BASE + "/index.html", { waitUntil: "domcontentloaded" });
  await click("#btn-enter");
  await click("#refuse-obit");
  t = await text(page);
  ok("ghost-obit", t.indexOf("不改") !== -1 || t.indexOf("只读") !== -1, t.slice(-200));
  await click("#refuse-wcs");
  t = await text(page);
  ok("ghost-print-wcs", t.indexOf("手印") !== -1, t.slice(-200));
  await click("#refuse-wgx");
  t = await text(page);
  ok("ghost-print-wgx", t.indexOf("手印") !== -1 && t.indexOf("吴桂香") !== -1, t.slice(-200));

  // 回访：放进黄叔
  await page.evaluate(() => { try { localStorage.removeItem("touqi-state"); } catch (e) {} });
  await page.goto(BASE + "/index.html", { waitUntil: "domcontentloaded" });
  await click("#btn-enter");
  await click("#btn-doc-flow");
  await click("#btn-admit"); // wcs
  await click("#btn-next");
  await click("#btn-doc-pencil");
  await click("#btn-admit"); // hs
  t = await text(page);
  ok("mind-drop", t.indexOf("神智") !== -1);
  const mind = await page.$eval("#dim-mind", (el) => el.textContent);
  ok("mind-value", mind === "3", mind);
  await click("#btn-next"); // 3
  if ((await page.$eval("#silhouette", (el) => el.textContent)).indexOf("无人") !== -1) await click("#btn-next");
  else {
    await click("#btn-omen");
    await click("#btn-doc-flow");
    await click("#btn-deny");
    await click("#btn-next");
  }
  // 4 桂香 deny → also revisit
  await click("#btn-omen");
  await click("#btn-doc-fang");
  await click("#btn-deny");
  await click("#btn-next"); // 5
  await click("#btn-omen");
  await click("#btn-doc-flow");
  await click("#btn-deny");
  await click("#btn-next"); // 6
  t = await text(page);
  ok("revisit", t.indexOf("第6晚") !== -1 && (t.indexOf("黄叔") !== -1 || t.indexOf("吴桂香") !== -1), t.slice(0, 500));

  // 刷新续班
  const nightBefore = await page.$eval("#night-mark", (el) => el.textContent);
  await page.reload({ waitUntil: "domcontentloaded" });
  t = await text(page);
  ok("refresh-restore", t.indexOf("第6晚") !== -1 && t.indexOf("进站") === -1, t.slice(0, 200) + " / " + nightBefore);

  // 来不及
  await page.evaluate(() => { try { localStorage.removeItem("touqi-state"); } catch (e) {} });
  await page.goto(BASE + "/index.html", { waitUntil: "domcontentloaded" });
  await click("#btn-enter");
  await page.evaluate(() => window.__TOUQI__.setEnergy(0));
  t = await text(page);
  ok("late", t.indexOf("来不及") !== -1 || t.indexOf("没交钥匙") !== -1, t.slice(-200));

  // 窄屏按钮高度
  await page.evaluate(() => { try { localStorage.removeItem("touqi-state"); } catch (e) {} });
  await page.setViewport({ width: 375, height: 667 });
  await page.goto(BASE + "/index.html", { waitUntil: "domcontentloaded" });
  await click("#btn-enter");
  const heights = await page.evaluate(() => {
    const ids = ["btn-admit", "btn-deny", "btn-key", "btn-doc-flow", "btn-next", "refuse-obit"];
    const out = {};
    ids.forEach((id) => {
      const el = document.getElementById(id);
      out[id] = el ? Math.round(el.getBoundingClientRect().height) : 0;
    });
    return out;
  });
  ok("btn-admit-h", heights["btn-admit"] >= 44, JSON.stringify(heights));
  ok("btn-key-h", heights["btn-key"] >= 44, JSON.stringify(heights));
  ok("btn-flow-h", heights["btn-doc-flow"] >= 44, JSON.stringify(heights));
  ok("btn-ghost-h", heights["refuse-obit"] >= 44, JSON.stringify(heights));
  const scroll = await page.evaluate(() => document.documentElement.scrollWidth <= window.innerWidth + 2);
  ok("narrow-no-xscroll", scroll);

  ok("no-pageerror", errors.length === 0, errors.join(" | "));
} catch (e) {
  console.log("THROW", e && e.stack ? e.stack : e);
  fails.push("throw");
} finally {
  await browser.close();
}

if (fails.length) {
  console.log("E2E_FAIL", fails.length, fails.join(" ;; "));
  process.exit(1);
}
console.log("E2E_OK");

File: 梁日班/英文版/qa/e2e.py
#!/usr/bin/env python3
"""Browser-click playtest for 梁日班. Buttons only; no engine.act()."""
import os
from playwright.sync_api import sync_playwright

SHOT = "/tmp/liang-gate/shots"
os.makedirs(SHOT, exist_ok=True)

BASE = os.environ.get("LIANG_BASE", "http://127.0.0.1:8766/index.html")
CHROME = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
fails = []
passes = []


def ok(name, cond, extra=""):
    if cond:
        print("PASS", name)
        passes.append(name)
    else:
        print("FAIL", name, extra or "")
        fails.append(name + ((" " + extra) if extra else ""))


def body(page):
    return page.evaluate("() => document.body.innerText")


def ending_text(page):
    return page.evaluate(
        """() => {
          const n = document.getElementById('ending-layer');
          return n ? n.innerText : '';
        }"""
    )


def state(page):
    return page.evaluate("() => window.__LIANG__ && window.__LIANG__.state")


def fresh(page):
    page.goto(BASE, wait_until="domcontentloaded")
    page.evaluate("() => { try { localStorage.removeItem('liang-state'); } catch (e) {} }")
    page.reload(wait_until="domcontentloaded")


def enter(page):
    page.click("#btn-enter")
    page.wait_for_selector("#desk.show")


def listen_day(page, paper_id, extras=None):
    page.click(paper_id)
    page.click("#btn-write")
    if extras:
        extras(page)
    st = state(page)
    if st and st.get("night") == 7 and page.locator("#btn-visit").is_visible():
        page.click("#btn-visit")
    page.click("#btn-submit")


def play_plan(page, papers, extras_on=None):
    enter(page)
    for n, pid in enumerate(papers, start=1):
        listen_day(page, pid, extras_on if n == extras_on else None)
        if n < len(papers):
            page.click("#btn-next")


def banned(text):
    words = ["主线", "关卡", "通关", "解锁", "线索", "结局", "源码", "本游戏", "唯一解", "进度条"]
    hit = [w for w in words if w in text]
    return hit


def run():
    with sync_playwright() as p:
        browser = p.chromium.launch(
            executable_path=CHROME,
            headless=True,
            args=["--no-sandbox", "--disable-gpu", "--no-proxy-server", "--proxy-bypass-list=*"],
        )
        page = browser.new_page(viewport={"width": 1280, "height": 800})
        page.set_default_timeout(8000)
        console = []
        pageerrors = []
        failed_req = []
        page.on("console", lambda m: console.append(f"{m.type}: {m.text}"))
        page.on("pageerror", lambda e: pageerrors.append(str(e)))
        page.on("requestfailed", lambda r: failed_req.append(f"{r.url} {r.failure}"))

        j = "#btn-jianli-line"
        r = "#btn-red"
        s = "#btn-safe"

        fresh(page)
        t = body(page)
        ok("title-who", "祁晚纸" in t)
        ok("title-task", "日落前" in t)
        ok("title-enter", page.locator("#btn-enter").is_visible())
        ok("title-lock", "今天只能听一份" in t)
        ok("title-fiction", "虚构" in t)
        ok("title-no-banned", not banned(t), str(banned(t)))
        page.screenshot(path=os.path.join(SHOT, "title.png"))

        page.focus("#btn-enter")
        page.keyboard.press("Enter")
        page.wait_for_selector("#desk.show")
        t = body(page)
        ok("keyboard-enter", "待听" in t and "写入主栏" in t)
        ok("boot-who", "祁晚纸" in t and "临聘文员" in t)
        ok("boot-hud-day", "第1日" in t)
        ok("boot-no-night-label", "第1晚" not in t and "下一晚" not in t)
        ok("boot-no-chengzu", "不成组" not in t)
        ok("boot-jianli-clickable", page.locator(j).is_visible())
        ok("boot-red-hidden-n1", not page.locator(r).is_visible())
        ok("boot-safe-hidden-n1", not page.locator(s).is_visible())
        ok("boot-pencil-hidden-n1", not page.locator("#paper-pencil").is_visible())
        ok("boot-kouxin-hidden-n1", not page.locator("#paper-kouxin").is_visible())
        ok("boot-visit-hidden-n1", not page.locator("#paper-visit").is_visible())
        ok("boot-no-banned", not banned(t), str(banned(t)))
        st = state(page)
        ok("boot-dims-4", st and st["dims"] == {"wage": 4, "mind": 4, "rep": 4, "mirror": 4})

        page.evaluate("() => document.getElementById('btn-red').click()")
        t = body(page)
        ok("n1-red-locked", "还没纸" in t)
        page.click(j)
        t = body(page)
        ok("first-pending", "待听" in t and "监理" in t)
        ok("first-limits", "不能证明吉时门外是谁" in t)
        page.click("#btn-write")
        page.click("#btn-submit")
        st = state(page)
        ok("n1-valid", st and st.get("slipSubmitted") and st.get("wage") == 36)
        ok("n1-no-ending", not st.get("ending"))

        page.click("#btn-next")
        st = state(page)
        t = body(page)
        ok("n2-pencil", page.locator("#paper-pencil").is_visible())
        ok("n2-red-still-hidden", not page.locator(r).is_visible())
        page.click("#btn-pencil")
        page.click("#btn-write")
        t = body(page)
        ok("pencil-reject", "不能当本日听份" in t or "主栏不收" in t)
        page.click("#btn-gossip")
        t = body(page)
        ok("gossip-idle", "闲话" in t)
        page.click("#refuse-contract")
        ok("refuse-contract", "不改合同原件" in body(page))
        page.click("#refuse-sign")
        ok("refuse-sign", "不替宅主签字" in body(page))
        page.click(j)
        page.click("#btn-write")
        page.click("#btn-submit")
        st = state(page)
        ok("n2-valid", st and st.get("slipSubmitted") and st.get("night") == 2)

        page.click("#btn-next")
        ok("n3-red-visible", page.locator(r).is_visible())
        ok("n3-safe-hidden", not page.locator(s).is_visible())

        page.reload(wait_until="domcontentloaded")
        ok("refresh-to-title", page.locator("#btn-enter").is_visible())

        # sunset: first click arms, second ends
        fresh(page)
        enter(page)
        page.click("#btn-sunset")
        st = state(page)
        t = body(page)
        ok("sunset-arm", not st.get("ending") and "再点一次" in t)
        page.click("#btn-sunset")
        st = state(page)
        e = ending_text(page)
        ok("end-late-sunset", st and st.get("ending") == "late" and "来不及" in (e + body(page)))
        ok("end-late-sunset-reason", "日落到了" in e or "缝里没有新条" in e)
        ok("end-late-in-view", page.evaluate("() => document.getElementById('ending-layer').getBoundingClientRect().top") < 700)
        page.screenshot(path=os.path.join(SHOT, "end-late.png"))

        # 4 empty → wage 0 → late
        fresh(page)
        enter(page)
        for n in range(4):
            page.click("#btn-submit")
            st = state(page)
            if st.get("ending"):
                break
            page.click("#btn-next")
        st = state(page)
        ok("end-late-wage0", st and st.get("ending") == "late" and st["dims"]["wage"] == 0)
        ok("end-late-wage-reason", "工钱格空了" in ending_text(page))

        # n7 empty after 6 valid → late (visit day no listen)
        fresh(page)
        play_plan(page, [j] * 6)
        page.click("#btn-next")
        page.click("#btn-submit")
        t = body(page)
        st = state(page)
        ok("n7-submit-needs-visit", not st.get("ending") and "先应他" in t)
        page.click("#btn-visit")
        page.click("#btn-submit")
        st = state(page)
        ok("end-late-n7-empty", st and st.get("ending") == "late")
        ok("end-late-n7-reason", "第七日主栏是空的" in ending_text(page))

        # dual on night 3
        fresh(page)
        play_plan(page, [j, j])
        page.click("#btn-next")
        page.click(j)
        page.click(r)
        t = body(page)
        st = state(page)
        ok("dual-reject-msg", "两份都听了" in t or "对不上" in t)
        ok("dual-mind", st and st["dims"]["mind"] == 3)
        page.click("#btn-submit")
        st = state(page)
        e = ending_text(page)
        ok("end-dual", st and st.get("ending") == "dual" and "双听" in e)
        ok("end-dual-reason", "两份打回后仍硬交" in e)
        page.screenshot(path=os.path.join(SHOT, "end-dual.png"))

        # dual recover
        fresh(page)
        play_plan(page, [j, j])
        page.click("#btn-next")
        page.click(j)
        page.click(r)
        page.click(j)
        page.click("#btn-write")
        page.click("#btn-submit")
        st = state(page)
        ok("dual-recover-valid", st and st.get("slipSubmitted") and st.get("wage") == 36 and not st.get("ending"))

        fresh(page)
        enter(page)
        disabled = page.evaluate("() => document.getElementById('btn-next').disabled")
        ok("next-disabled-before-submit", disabled is True)
        page.click("#btn-write")
        ok("write-need-pending", "先点一份" in body(page))

        # ji: 7×监理
        fresh(page)
        play_plan(page, [j] * 7)
        st = state(page)
        e = ending_text(page)
        top = page.evaluate("() => document.getElementById('ending-layer').getBoundingClientRect().top")
        ok("end-ji", st and st.get("ending") == "ji" and "寄梁" in e)
        ok("end-ji-reason", "七日你都听监理" in e)
        ok("end-ji-in-view", top < 700, str(top))
        ok("end-ji-wage36", st and st.get("wage") == 36)
        page.screenshot(path=os.path.join(SHOT, "end-ji.png"))

        # sheng: n1-2 监理, n3-7 红纸
        fresh(page)
        play_plan(page, [j, j, r, r, r, r, r])
        st = state(page)
        e = ending_text(page)
        ok("end-sheng", st and st.get("ending") == "sheng" and "升梁" in e)
        ok("end-sheng-reason", "听东家" in e and "回访" in e)
        ok("end-sheng-rep-down", st and st["dims"]["rep"] < 4)

        # qing: n1-3 监理, n4-7 安全单
        fresh(page)
        play_plan(page, [j, j, j, s, s, s, s])
        st = state(page)
        e = ending_text(page)
        ok("end-qing", st and st.get("ending") == "qing" and "清场" in e)
        ok("end-qing-reason", "听安全单" in e and "回访" in e)
        ok("end-qing-mirror-down", st and st["dims"]["mirror"] < 4)

        # n6 kouxin burden
        fresh(page)
        play_plan(page, [j] * 5)
        page.click("#btn-next")
        st = state(page)
        t = body(page)
        ok("n6-now", st and st.get("night") == 6)
        ok("n6-hud-day", "第6日" in t and "第6晚" not in t)
        ok("n6-kouxin-visible", page.locator("#paper-kouxin").is_visible())
        page.click("#btn-kouxin")
        t = body(page)
        ok("n6-kouxin-limits", "不能当本日听份" in t and "主栏能收两格" in t)
        page.click("#btn-write")
        t = body(page)
        ok("n6-kouxin-not-main", "主栏只有一格" in t and "不收" in t)
        page.click(j)
        page.click("#btn-write")
        page.click("#btn-submit")
        st = state(page)
        ok("n6-after-kouxin-valid", st and st.get("slipSubmitted") and st.get("wage") == 36)

        # trust-red: last night red + 把红纸当全真
        fresh(page)
        enter(page)
        seq = [j, j, r, r, r, r, r]
        for n, pid in enumerate(seq, start=1):
            page.click(pid)
            page.click("#btn-write")
            if n == 7:
                page.click("#btn-trust-red")
                page.click("#btn-visit")
            page.click("#btn-submit")
            if n < 7:
                page.click("#btn-next")
        st = state(page)
        e = ending_text(page)
        ok("end-trust-red", st and st.get("ending") == "trust-red" and "信红纸" in e)
        ok("end-trust-reason", "把红纸当全真" in e)
        page.screenshot(path=os.path.join(SHOT, "end-trust-red.png"))

        # layers
        fresh(page)
        enter(page)
        for n in range(1, 8):
            st = state(page)
            night = st["night"]
            ok(f"layer-red-n{night}", page.locator("#paper-red-block").is_visible() == (night >= 3))
            ok(f"layer-safe-n{night}", page.locator("#paper-safe-block").is_visible() == (night >= 4))
            ok(f"layer-pencil-n{night}", page.locator("#paper-pencil").is_visible() == (night >= 2))
            ok(f"layer-print-n{night}", page.locator("#paper-print").is_visible() == (night >= 5))
            ok(f"layer-kouxin-n{night}", page.locator("#paper-kouxin").is_visible() == (night >= 6))
            ok(f"layer-visit-n{night}", page.locator("#paper-visit").is_visible() == (night >= 7))
            listen_day(page, j)
            if n < 7:
                page.click("#btn-next")

        # beam / crack
        fresh(page)
        enter(page)
        page.click("#btn-beam")
        ok("beam-no-listen", "影子" in body(page) or "寄梁绳位" in body(page))
        listen_day(page, j)
        for _ in range(3):
            page.click("#btn-next")
            listen_day(page, j)
        page.click("#btn-next")
        st = state(page)
        ok("n5-now", st and st["night"] == 5)
        page.click("#btn-crack")
        ok("crack-old-name", "旧宅主名" in body(page) or "不是方守成" in body(page))

        page.set_viewport_size({"width": 320, "height": 640})
        fresh(page)
        enter(page)
        boxes = page.evaluate(
            """() => {
              const ids = ['btn-jianli-line','btn-write','btn-submit','btn-sunset'];
              return ids.map(id => {
                const el = document.getElementById(id);
                const r = el.getBoundingClientRect();
                return {id, w: r.width, h: r.height, vis: r.height>0 && r.width>0};
              });
            }"""
        )
        ok("narrow-btns-visible", all(b["vis"] for b in boxes), str(boxes))
        ok("narrow-tap-height", all(b["h"] >= 32 for b in boxes), str(boxes))
        page.click(j)
        page.click("#btn-write")
        page.click("#btn-submit")
        st = state(page)
        ok("narrow-core-loop", st and st.get("slipSubmitted") and st.get("wage") == 36)

        page.set_viewport_size({"width": 1280, "height": 800})
        fresh(page)
        page.focus("#btn-enter")
        page.keyboard.press("Enter")
        page.wait_for_selector("#desk.show")
        page.focus(j)
        page.keyboard.press("Enter")
        page.focus("#btn-write")
        page.keyboard.press("Enter")
        page.focus("#btn-submit")
        page.keyboard.press("Enter")
        st = state(page)
        ok("keyboard-first-group", st and st.get("slipSubmitted") and st.get("listenSource") == "jianli")

        missing = [u for u in failed_req if "jpeg/" in u or u.endswith(".js") or u.endswith(".css")]
        ok("no-failed-assets", not missing, str(missing)[:400])
        ok("no-pageerror", not pageerrors, str(pageerrors)[:400])
        err_console = [c for c in console if c.startswith("error") and "favicon" not in c]
        ok("no-console-error", not err_console, str(err_console)[:400])

        browser.close()

    print("---")
    print("PASS_COUNT", len(passes))
    print("FAIL_COUNT", len(fails))
    for f in fails:
        print("FAIL_ITEM", f)
    return 0 if not fails else 1


if __name__ == "__main__":
    raise SystemExit(run())
